### ISSUE TITLE
[macOs] Tap gestures not working properly with magic mouse

### ISSUE THREAD

**Issue by knopp** [flutter/flutter#120788]
Code:

```dart
      body: GestureDetector(
          behavior: HitTestBehavior.opaque,
          onPanStart: (details) => {
            print('onPanStart'),
          },
          child: GestureDetector(
            behavior: HitTestBehavior.opaque,
            onTapUp: (e) {
              print('onTapUp');
            },
          ),
        ),
      ),
```

To reproduce:
1. Rest both fingers on the mouse
2. Scroll with right finger

After that clicking with left finger will not report the *tap up* gesture until user lifts the left finger.

I think the problem here is that in this case macOS reports the pan zoom event in parallel with mouse down/up events, which is something the framework is not prepared to handle. The pan gesture recognizer always wins until the gesture event with NSEventPhaseEnded is dispatched, which happens after left finger lift.

cc @moffatman

**Comment by moffatman** [comment]
I don't think I can help here, I don't have a magic mouse to reproduce it. Doing the same sequence on trackpad works as expected.

**Comment by moffatman** [comment]
Does it work without `behavior: HitTestBehavior.opaque`? Also from your description it sounds like the framework is accurately reacting to the macOS sequence? What's broken specifically, can you show {macOS state changes, current framework state changes, desired framework state changes}?

**Comment by knopp** [comment]
Hit test behavior makes no difference here. I just put it in example because there is no child widget.

The framework might be accurately reacting to engine events, but the end result is incorrect and different from native macOS applications.

In native applications it is possible to initiate scroll with right finger on magic mouse and left finger will continue registering clicks (without having to lift the left finger from the mouse)

In flutter once scroll is initiated, the left finger will not register clicks anymore until it is lifted from the mouse. This makes the application seems unresponsive (since user should not be require to lift left finger from the mouse after scroll).

Part of the problem seems to be in the framework: Once the `PointerPanZoomStartEvent` is received (pan in progress) then PointerUpEvent will not trigger the tap recognizer though they have different pointer ids.

---

That said, looking at the events, there is a bit of weirdness. After scrolling with right finger, once the scrolling is done, there are additional pan events:

PointerPanZoomEndEvent // dispatched after lifting right finger
PointerPanZoomStartEvent // this is dispached after the right finger was already lifted
PointerPanZoomUpdateEvent // dispatched after moving left finger (right finger no longer on mouse)
PointerPanZoomEndEvent // dispatched after removing left finger from mouse - after this clicking will work

I don't know why these events are generated by cocoa. But if we could filter them out in the embedder that should fix the situation.

**Comment by moffatman** [comment]
Can you print the framework events collected via `Listener` as well?

**Comment by knopp** [comment]
You are right. That's not weirdness, that's expected behavior. The problem is that with native applications, in this state (continued scrolling) clicking works as expected. But with Flutter the tap gesture recognizer will not be invoked, because the pan recognizer already won and is in progress.

Here is the whole sequence of events:
```
Event PointerAddedEvent#2197d(position: Offset(315.6, 223.5)), pointer 0  // mouse entered window
Event PointerPanZoomStartEvent#0d986(position: Offset(323.4, 107.6)), pointer 6 // started scrolling with right finger, left finger resting on mouse
onPanStart
Event PointerPanZoomUpdateEvent#04464(position: Offset(323.4, 107.6)), pointer 6
Event PointerPanZoomUpdateEvent#baccb(position: Offset(323.4, 107.6)), pointer 6
Event PointerPanZoomUpdateEvent#69202(position: Offset(323.4, 107.6)), pointer 6
Event PointerPanZoomEndEvent#cb2e9(position: Offset(323.4, 107.6)), pointer 6 // lifted right finger
Event PointerPanZoomStartEvent#7bb5b(position: Offset(323.4, 107.6)), pointer 7 // scrolling continues with left finger
onPanStart
Event PointerPanZoomUpdateEvent#02a37(position: Offset(323.4, 107.6)), pointer 7 // scrolling with left finger
Event PointerPanZoomUpdateEvent#6bdd1(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#10758(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#372e3(position: Offset(323.4, 107.6)), pointer 7
Event PointerDownEvent#1d68c(position: Offset(323.4, 107.6)), pointer 8 // pressing down left finger
Event PointerPanZoomUpdateEvent#76188(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#e830d(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#303c0(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#46dcd(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#41b89(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#c6ceb(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#e6f71(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#9dc9f(position: Offset(323.4, 107.6)), pointer 7
Event PointerPanZoomUpdateEvent#b597b(position: Offset(323.4, 107.6)), pointer 7
Event PointerUpEvent#e71b7(position: Offset(323.4, 107.6)), pointer 8 // up left finger - does not generate tap recognizer because pan is in progress (despite different pointer number)
Event PointerPanZoomEndEvent#328e6(position: Offset(323.4, 107.6)), pointer 7 // lifted left finger, clicks will work
Event PointerRemovedEvent#f8abe(position: Offset(395.0, 241.0)), pointer 0 
```

**Comment by knopp** [comment]
Btw. I think I overcomplicated this with the right finger. You can just scroll with just one finger. Normally the clicks should be recognized during scrolling/panning (macOS behavior), but in flutter they are not.

**Comment by knopp** [comment]
The fact that mouse down / up events have different pointer numbers but don't trigger tap gesture recognizer seems weird to me.

**Comment by moffatman** [comment]
The pan gesture recognizer will always claim new pointers if it's already active. Which makes a lot of sense on touch, but maybe less on desktop. But this multiple-simultaneous-input scenario on desktop is rare so it hasn't come up yet.

Maybe we should change `DragGestureRecognizer::_addPointer` to not immediately claim pointer if the new device-kind is a mouse and the previous device-kind is not a mouse? Same thing also would be needed in `ScaleGestureRecognizer::_advanceStateMachine`.

**Comment by knopp** [comment]
You're right. Only resolving the gesture inside `DragGestureRecognizer::_addPointer` when `event.kind != PointerDeviceKind.mouse` seems to work.

**Comment by moffatman** [comment]
These recognizers don't have a way to be partially-accepted. So if we make this change, the mouse-cursor drag can never join in with the pan/scale, even if you start moving the mouse itself. I don't think that will break anything, though. 

Not sure if you can test the native macOS behavior here as in general you can't scroll by dragging the mouse, so you'll have to find some interaction where both trackpad and mouse-drag do the same thing.

**Comment by knopp** [comment]
So `_addPointer` doesn't seem to cover all cases. Because the recognizer will still track the pointer (isPointerAllowed returns true) from mouse down and it may get resolved by the arena.

It seems that a solution might be to remember initial pointer in the `DragGestureRecognizer` and then if `kind == mouse` reject all other pointers. But right now I'm not sure what the sideffects will be. It's a progress though :)

**Comment by moffatman** [comment]
We can override isPointerAllowed then?

**Comment by Piinks** [comment]
Hi @knopp can you provide a complete runnable code sample that reproduces the issue?

**Comment by knopp** [comment]
> Hi @knopp can you provide a complete runnable code sample that reproduces the issue?

The code above is the complete code. To reproduce simply scroll with the magic mouse. After scrolling the `onTapUp` recognizer will not be triggered until the finger list lifted off the mouse.

I updated the original description as this can be easily reproducible with one finger.

I think the fix might be adding another condition in `DragGestureRecognizer.isPointerAllowed`. Similar to how we do not allow adding new buttons during drag

https://github.com/flutter/flutter/blob/d6de6bc681063c1ade090d71cb3010101e5b58dd/packages/flutter/lib/src/gestures/monodrag.dart#L260-L264

we should not allow adding new touches when device is mouse:

```dart
    if (_initialPointer != null &&
        event.kind == PointerDeviceKind.mouse &&
        event.pointer != _initialPointer) {
      return false;
    }
```

Because with magic mouse it is possible for one finger to control pan gesture while another finger is dragging.
